### DISCUSSION TITLE
Display Spotlight as wireframe

### DISCUSSION THREAD

**Discussion by wes-nz** [google/filament#4694] [Q&A]
I am trying to find a way to show the boundaries of my spotlights. In the same way the filament app is doing it for the camera.

![Screen Shot 2021-10-02 at 5 32 00 PM](https://user-images.githubusercontent.com/2945380/135703583-58db9a3d-dd8d-4faa-96d9-6c93ed4eafb1.png)

I have been getting inspiration from [`FilamentApp.cpp`](https://github.com/google/filament/blob/3d51f9ab2346c8c5c64ab94fb80ddd3b3f401667/libs/filamentapp/src/FilamentApp.cpp) and [`Cube.cpp`](https://github.com/google/filament/blob/24b88219fa6148b8004f230b377f163e6f184d65/libs/filamentapp/src/Cube.cpp#L77-L84).  As far as I am aware, it is not possible to use a similar solution because lights dont have any bounding box.

I assume the second best solution is to have a cone matching the spotlight configuration. Is there any other approaches you would recommend?

Thanks.

**Comment by romainguy** [comment]
You could approximate the spotlight come with a frustum similar to the cameras but a cone wireframe would be better. You can generate one procedurally fairly easily, there are a bunch of articles online on this. For instance: https://gist.github.com/andresfelipemendez/ef4f832084822bcf400600879c3b02e1

**Reply by wes-nz** [reply]
Thanks, that will be useful! There is no way to get a model matrix from a spotlight right? (Similar to `getModelMatrix` for a camera).

**Reply by romainguy** [reply]
The model matrix of a spotlight is provided by the transform manager for entities (lights, renderables, etc.).

**Reply by wes-nz** [reply]
Oh right of course. Looking at the code will be easier. I am stuck converting the following but for a light entity 

```cpp
void Cube::mapFrustum(filament::Engine& engine, Camera const* camera) {
    // the Camera far plane is at infinity, but we want it closer for display
    const mat4 vm(camera->getModelMatrix());
    mat4 p(vm * inverse(camera->getCullingProjectionMatrix()));
    return mapFrustum(engine, p);
}
```

becomes (pseudo-code)
```cpp
void Cube::mapFrustum(filament::Engine& engine, Entity spotlight) {
    auto& tcm = engine.getTransformManager();
    const mat4 vm(tcm.getWorldTransform(tcm.getInstance(spotLight));
    mat4 p(vm * .................... ); // ??
    return mapFrustum(engine, p);
}
```
What am I supposed to use for ` inverse(camera->getCullingProjectionMatrix())`?
Thanks.
